src/allmydata/web/common.py

```diff
@@ -479,31 +476,230 @@ def render(self, req):
+        )
+        log.err("Request for {!r} handled by {!r} returned unusable {!r}".format(
+            request.uri,
+            fullyQualifiedName(render),
+            result,
+        ))
+        request.setResponseCode(http.INTERNAL_SERVER_ERROR)
+        _finish(b"Internal Server Error", render, request)
+
+
+def _renderHTTP_exception(request, failure):
+    try:
+        text, code = humanize_failure(failure)
+    except:
+        log.msg("exception in humanize_failure")
+        log.msg("argument was %s" % (failure,))
+        log.err()
+        text = str(failure)
+        code = None
+
+    if code is not None:
+        return _renderHTTP_exception_simple(request, text, code)
+
+    if failure.check(UnsupportedMethod):
+        # twisted.web.server.Request.render() has support for transforming
+        # this into an appropriate 501 NOT_IMPLEMENTED or 405 NOT_ALLOWED
+        # return code, but nevow does not.
+        method = request.method
+        return _renderHTTP_exception_simple(
+            request,
+            "I don't know how to treat a %s request." % (method,),
+            http.NOT_IMPLEMENTED,
+        )
+
+    accept = request.getHeader("accept")
+    if not accept:
+        accept = "*/*"
+    if "*/*" in accept or "text/*" in accept or "text/html" in accept:
+        request.setResponseCode(http.INTERNAL_SERVER_ERROR)
+        return template.renderElement(
+            request,
+            tags.html(
+                tags.head(
+                    tags.title(u"Exception"),
+                ),
+                tags.body(
+                    FailureElement(failure),
+                ),
+            ),
+        )
+
+    # use plain text
+    traceback = failure.getTraceback()
+    return _renderHTTP_exception_simple(
+        request,
+        traceback,
+        http.INTERNAL_SERVER_ERROR,
+    )
+
+
+def _renderHTTP_exception_simple(request, text, code):
+    request.setResponseCode(code)
+    request.setHeader("content-type", "text/plain;charset=utf-8")
+    if isinstance(text, unicode):
+        text = text.encode("utf-8")
+    request.setHeader("content-length", b"%d" % len(text))
+    return text
+
+
+def handle_when_done(req, d):
+    when_done = get_arg(req, "when_done", None)
+    if when_done:
+        d.addCallback(lambda res: DecodedURL.from_text(when_done.decode("utf-8")))
+    return d
+
+
+def url_for_string(req, url_string):
+    """
+    Construct a universal URL using the given URL string.
+
+    :param IRequest req: The request being served.  If ``redir_to`` is not
+        absolute then this is used to determine the net location of this
+        server and the resulting URL is made to point at it.
+
+    :param bytes url_string: A byte string giving a universal or absolute URL.
+
+    :return DecodedURL: An absolute URL based on this server's net location
+        and the given URL string.
+    """
+    url = DecodedURL.from_text(url_string.decode("utf-8"))
```

Comment:
I guess that's it's job.  If anything, perhaps this is safer than what was here before since the URL must now undergo a roundtrip through DecodedURL instead of just being spit back out as-is.  If a client is tricked into including a malicious URL that is somehow sneakily malformed here, the old implementation would spit it back at them but the new implementation will raise an exception or normalize it to something safe.

I guess I can think of some kind of response splitting attack where the URL has embedded newlines and an HTTP response.  OTOH if an attacker can control this URL the user may already have lost, independent of any such tricks (eg sometimes this allows the attacker to get the newly created cap).